crates/red_knot_python_semantic/src/types.rs

```diff
@@ -463,6 +463,140 @@ impl<'db> Type<'db> {
         self == other
     }
 
+    /// Return true if this type and `other` have no common elements.
+    ///
+    /// Note: This function aims to have no false positives, but might return
+    /// wrong `false` answers in some cases.
+    pub(crate) fn is_disjoint_from(self, db: &'db dyn Db, other: Type<'db>) -> bool {
+        match (self, other) {
```

Comment:
There are 18 enum variants in `Type`, so 18 × 18 = 324 cases to cover. `is_disjoint_from` is symmetric, so in principle it's only one triangular part of the matrix, but that still leaves 171 cases. Not sure this is the best way to structure them...